lib/driver.js

```diff
@@ -185,37 +185,43 @@ class XCUITestDriver extends BaseDriver {
 
   async determineDevice () {
     let realDevice, udid;
-    let sim;
+    let device;
     if (this.opts.udid) {
       if (await simExists(this.opts.udid)) {
         this.lifecycleData.createSim = false;
         udid = this.opts.udid;
-        sim = await getSimulator(udid);
+        device = await getSimulator(udid);
         realDevice = false;
       } else {
-        //TODO check for real device
-        throw new errors.SessionNotCreatedError(`no simulator found with udid ${this.opts.udid}`);
+        if (this.opts.udid.toLowerCase() === 'auto') {
+          this.opts.udid = await detectUdid();
+        }
+        udid = this.opts.udid;
+        realDevice = true;
+
+        // TODO: decide what object this should be
+        device = {udid};
       }
     } else {
-      sim = await getExistingSim(this.opts.deviceName, this.opts.platformVersion);
-      if (sim) {
+      device = await getExistingSim(this.opts.deviceName, this.opts.platformVersion);
+      if (device) {
         if (this.opts.reset || this.opts.fullReset) {
-          await sim.clean();
-          await sim.shutdown();
+          await device.clean();
+          await device.shutdown();
         }
         this.lifecycleData.createSim = false;
-        udid = sim.udid;
+        udid = device.udid;
         realDevice = false;
       } else {
         // no udid: create a new sim
         log.info(`udid not provided, using desired caps to create a new sim`);
-        sim = await this.createSim();
-        udid = sim.udid;
+        device = await this.createSim();
+        udid = device.udid;
         realDevice = false;
       }
     }
 
-    return { sim, udid, realDevice };
+    return { device, udid, realDevice };
```

Comment:
just for readability it could be nice to flatten this section out:

```js
if (this.opts.udid && await simExists(this.opts.udid)) {
  ...
  return {device, udid, realDevice};
}

if (this.opts.udid) {
  ...
  return {device, udid, realDevice};
}

device = await getExistingSim(this.opts.deviceName, this.opts.platformVersion);

if (device) {
  ...
  return {device, udid, realDevice};
}

log.info(`udid not provided...`);
...
return {device, udid, realDevice};
```

It's up to you; there's a little more repetition in the `return`s, but I find multiple embedded conditionals hard to follow especially on small screen sizes. you have to keep more state in your head.